search-app/src/cmr/search/services/parameters/parameter_validation.clj

```diff
@@ -366,41 +366,70 @@
+          [error-msg]
           (reduce
             (fn [errors param]
-              (if-not (some #{param} (nf/get-subfield-names :science-keywords))
-                (conj errors (format "parameter [%s] is not a valid science keyword search term."
-                                     (name param)))
+              (if-not (some #{param} (nf/get-subfield-names field))
+                (conj errors (format "parameter [%s] is not a valid [%s] search term."
+                                     (name param)
+                                     (csk/->snake_case_string field)))
                 errors))
             []
             (mapcat keys values))))
-      [(msg/science-keyword-invalid-format-msg)])))
+      [error-msg])))
+
+(defn science-keywords-validation-for-field
+  "Performs science keywords subfield validation."
+  [field concept-type params]
+  (nested-field-validation-for-subfield field concept-type params
+                                        (msg/science-keyword-invalid-format-msg)))
 
 (defn variables-validation
   "Validates the variables-h search parameters are in the format of e.g.
-   variables-h[0][measurement]=vaule."
+   variables-h[0][measurement]=value."
   [concept-type params]
-  (when-let [variables (get params :variables-h)]
-    (if (map? variables)
-      (let [values (vals variables)]
-        (if (some #(not (map? %)) values)
-          [(msg/variable-invalid-format-msg)]
-          (reduce
-            (fn [errors param]
-              (if-not (some #{param} (nf/get-subfield-names :variables))
-                (conj errors (format "parameter [%s] is not a valid variable search term."
-                                     (name param)))
-                errors))
-            []
-            (mapcat keys values))))
-      [(msg/variable-invalid-format-msg)])))
+  (nested-field-validation-for-subfield :variables-h concept-type params
+                                        (msg/variable-invalid-format-msg)))
+
+(defn temporal-facets-subfields-validation
+  "Performs temporal facets subfield validation."
+  [concept-type params]
+  (nested-field-validation-for-subfield :temporal-facet concept-type params
+                                        (msg/temporal-facets-invalid-format-msg)))
+
+(defn valid-year?
+  "Returns true if the provided value is a valid year and false otherwise."
+  [year]
+  (let [parsed-year (try
+                      (Long. year)
+                      (catch Exception e
+                        nil))]
+    (and (not (nil? parsed-year))
+         (<= 1 parsed-year 9999))))
+
+(defn temporal-facet-year-validation
+  "Validates that the years provided in all temporal-facet parameters are valid."
+  [concept-type params]
+  (when-let [param-values (:temporal-facet params)]
+    (when (map? param-values)
+      (let [temporal-facet-maps (vals param-values)
+            years (keep :year temporal-facet-maps)]
+        (reduce
+          (fn [errors year]
```

Comment:
We should think about housing anonymous functions like this in their own namespace (this PR isn't the place for it). Typically, validation libraries have a collection of predicate-like functions in one place of the codebase; this is essential for reusability (obviously) but also, with good function names, this makes reading the validators much nicer. Again, something to think about.